src/Microsoft.Diagnostics.EventFlow.Outputs.ElasticSearch/ElasticSearchOutput.cs

```diff
@@ -378,11 +382,77 @@ private async Task EnsureIndexExists(string indexName, ElasticClient esClient)
             }
         }
 
-        private string GetIndexName(ElasticSearchConnectionData connectionData)
+        protected internal string GetIndexName(ElasticSearchConnectionData connectionData)
+        {
+            string dateString = "";
+            try
+            {
+                dateString = ConvertIndexFormat(connectionData.Configuration.IndexFormat);
+            }
+            catch (Exception ex)
+            {
+                string errorMessage = $"{nameof(ElasticSearchOutput)}: indexFormat could not parse as date: {ex.Message}{Environment.NewLine}" +
+                                      $"Invalid IndexFormat value: {connectionData.Configuration.IndexFormat}{Environment.NewLine}" +
+                                      $"Falling back to default value: {ElasticSearchOutputConfiguration.DefaultIndexFormat}";
+                // Changing in connectionData.Configuration so we don't report this every log message
+                connectionData.Configuration.IndexFormat = ElasticSearchOutputConfiguration.DefaultIndexFormat;
+                healthReporter.ReportWarning(errorMessage);
+                dateString = string.Format($"{{0:{connectionData.Configuration.IndexFormat}}}", DateTimeOffset.UtcNow);
+            }
+
+            return connectionData.Configuration.IndexNamePrefix + dateString;
+        }
+
+        public Func<DateTimeOffset> Now { get; set; } = DefaultNow;
+        public static DateTimeOffset DefaultNow() { return DateTimeOffset.UtcNow; }
+
+        protected internal string ConvertIndexFormat(string dateString)
+        {
+            var index = dateString.IndexOf('q');
+            if (index > -1)
+                dateString = dateString.Substring(0, index) + GetQuarterOfYear() + dateString.Substring(index + 1);
+            index = dateString.IndexOf('w');
+            if (index > -1)
+                dateString = dateString.Substring(0, index) + GetWeekOfMonth() + dateString.Substring(index + 1);
+            else
+            {
+                index = dateString.IndexOf("WW", StringComparison.Ordinal);
+                if (index > -1)
+                    dateString = dateString.Substring(0, index) + GetWeekOfYear(true) + dateString.Substring(index + 2);
+                else
+                {
+                    index = dateString.IndexOf('W');
+                    if (index > -1)
+                        dateString = dateString.Substring(0, index) + GetWeekOfYear(false) +
+                                     dateString.Substring(index + 1);
+                }
+            }
+
+            return string.Format($"{{0:{dateString}}}", Now());
+        }
+
+        private string GetQuarterOfYear()
+        {
+            return "q"+ (((Now().Month - 1) / 3) + 1);
+        }
+
+        private string GetWeekOfMonth()
+        {
+            DateTimeOffset date = Now();
```

Comment:
This and the next variable do not seem to be used to compute the answer?

Reply:
Refactor bug. I used to have the date in the parameter list and removed it, not noticing that it was actually needed because of how WeekOfMonth is calculated.